src/setup/node.rs

```diff
@@ -131,9 +152,89 @@ impl Node {
-            listener.accept().await.unwrap();
+        self.perform_initial_action(listener).await;
+    }
+
+    async fn perform_initial_action(&self, listener: Option<TcpListener>) {
+        match self.config.initial_action {
+            Action::None => {}
+            Action::WaitForConnection(_) => {
+                listener.unwrap().accept().await.unwrap();
+            }
+            Action::SeedWithGenesisBlocks(_) if self.meta.kind == NodeKind::Zebra => {
+                // not supported for zebra yet, so we just wait for connection at least
+                listener.unwrap().accept().await.unwrap();
+            }
+            Action::SeedWithGenesisBlocks(_) => {
+                use crate::protocol::message::Message;
+                let mut stream = respond_to_handshake(listener.unwrap()).await.unwrap();
+
+                let filter = MessageFilter::with_all_auto_reply()
+                    .with_getheaders_filter(Filter::Disabled)
+                    .with_getdata_filter(Filter::Disabled);
+
+                let block_0 = Box::new(Block::genesis_0());
+                let block_1 = Box::new(Block::genesis_1());
+                let block_2 = Box::new(Block::genesis_2());
+
+                // respond to GetHeaders(Block[0])
+                match filter.read_from_stream(&mut stream).await.unwrap() {
+                    Message::GetHeaders(locations) => {
+                        // The request should be from the genesis hash onwards,
+                        // i.e. locator_hash = [genesis.hash], stop_hash = [0]
+                        assert_eq!(
+                            locations.block_locator_hashes,
+                            vec![block_0.double_sha256().unwrap()]
+                        );
+                        assert_eq!(locations.hash_stop, Hash::zeroed());
+
+                        // Reply with headers for blocks [1,2]
+                        Message::Headers(Headers::new(vec![
+                            block_1.header.clone(),
+                            block_2.header.clone(),
+                        ]))
+                        .write_to_stream(&mut stream)
+                        .await
+                        .unwrap();
+                    }
+                    msg => panic!("Expected GetHeaders but got: {:?}", msg),
+                }
+
+                // respond to GetData(inv) for blocks([1,2])
+                match filter.read_from_stream(&mut stream).await.unwrap() {
+                    Message::GetData(inv) => {
+                        // The request must be for blocks [1,2]
+                        let expected = Inv::new(vec![
+                            InvHash::new(ObjectKind::Block, block_1.double_sha256().unwrap()),
+                            InvHash::new(ObjectKind::Block, block_2.double_sha256().unwrap()),
+                        ]);
+
+                        assert_eq!(inv, expected);
+
+                        // send blocks 1 and 2
```

Comment:
Should the genesis block be sent as well? I'm not sure nodes come pre-seeded with the genesis block?

Reply:
I was wondering about that, but the node never actually requests it? Initially I expected `inv` to contain a request for all three blocks, but it only requests the last two.

Reply:
Per discussion, seems like it is ignored by the nodes when sent: Nodes must be pre-seeded with the block.